metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala

```diff
@@ -0,0 +1,150 @@
+package scala.meta.internal.metals
+
+import scala.meta._
+import org.eclipse.lsp4j.SymbolKind
+import org.eclipse.lsp4j.DocumentSymbol
+import MetalsEnrichments._
+import scala.collection.concurrent.TrieMap
+
+/**
+ *  Retrieves all the symbols defined in a document
+ *
+ *  If the document doesn't parse, we fall back to the latest
+ *  known snapshot of the document, if present
+ *
+ *  To avoid the snapshots to grow undefinitely, we only store
+ *  the ones for the most recently requested documents
+ */
+class DocumentSymbolProvider(buffers: Buffers) {
+
+  def empty: List[DocumentSymbol] = Nil
+
+  def documentSymbols(
+      path: AbsolutePath
+  ): List[DocumentSymbol] = {
+    path
+      .toInputFromBuffers(buffers)
+      .parse[Source]
+      .toOption
+      .map { source =>
+        val result = new SymbolTraverser(path.toString).apply(source)
+        snapshots.put(path, result)
+        result
+      }
+      .orElse(snapshots.get(path))
+      .getOrElse(empty)
+  }
+
+  def discardSnapshot(path: AbsolutePath): Unit =
+    snapshots.remove(path)
+
+  private val snapshots = new TrieMap[AbsolutePath, List[DocumentSymbol]]()
```

Comment:
nit, `TrieMap.empty` for consistency with rest of the codebase

Reply:
nit 2, I generally try to declare private mutable state  at the top of the class.